src/librustc/ty/constness.rs

```diff
@@ -68,14 +69,54 @@ impl<'tcx> TyCtxt<'tcx, 'tcx> {
 
 
 pub fn provide<'tcx>(providers: &mut Providers<'tcx>) {
-    /// only checks whether the function has a `const` modifier
+    fn is_const_evaluatable(tcx: TyCtxt<'tcx, 'tcx>, def_id: DefId) -> bool {
```

Comment:
I would not have this as a separate function, but rather use it before the usage of `FnLikeNode`, so intrinsics' definitions aren't checked for `const fn` (as they syntactically cannot be `const fn`).

Reply:
> (as they syntactically cannot be const fn).

Shouldn't we just allow them to be that syntactically?

Reply:
Didn't you say you tried that once and it did not work?

Also, that would allow nightly users to declare intrinsics as `const`. There might be ways to cause unsoundness there. Wouldn't that be a problem?

Reply:
> Didn't you say you tried that once and it did not work?

It would be a lot of work

> Also, that would allow nightly users to declare intrinsics as const. There might be ways to cause unsoundness there. Wouldn't that be a problem?

Well... if there's an implementation in the miri-engine, then they can do whatever unstable fun they want I guess?

Reply:
Yes. And that might subvert properties that `const fn` otherwise have.

Reply:
I only consider this a problem in libstd. If you are using nightly features you can already brick properties that we aren't sure about yet, so...

Reply:
One of these days we should change intrinsics so they don't abuse `extern "..." {...}` anymore and then we can have `const fn transmute` or w/e. As for not allowing `const fn` on intrinsics that shouldn't be, I suppose that should go in `rustc_typeck`'s list of intrinsic signatures.

Reply:
We could just move to making them lang items and having libcore implement them by recursing on themselves and then have the compiler overwrite them. Then they'd be regular functions (and we used this trick on one intrinsic already: https://github.com/rust-lang/rust/blob/605ea9d05c48957a291eec11eb7339788c3140ed/src/libcore/ptr/mod.rs#L197)

Reply:
That would make it harder to figure out that a function is an intrinsic. Also `real_drop_in_place` actually codegens to real functions, while intrinsics always inline their code at the caller site. Another difference between intrinsics and (lang item) functions is that intrinsics cant be turned into function pointers, while (lang item) functions can be.

Reply:
@oli-obk *sigh* we don't have to re-use lang items, we can have a different mechanism (but also attribute-based) for them. I've been meaning to do this for ages but @nikomatsakis wanted an RFC which I never got around to writing, *oops*.